filters/auth/jwt_validation.go

```diff
@@ -0,0 +1,171 @@
+		Timeout: timeout,
+		Tracer:  opentracing.NoopTracer{},
+	})
+}
+
+func JwtValidationWithOptions(
+	create func(time.Duration) filters.Spec,
+	o TokenintrospectionOptions,
+) filters.Spec {
+	s := create(o.Timeout)
+	ts, ok := s.(*jwtValidationSpec)
+	if !ok {
+		return s
+	}
+
+	ts.options = o
+	return ts
+}
+
+func (s *jwtValidationSpec) Name() string {
+	return JwtValidationName
+}
+
+func (s *jwtValidationSpec) CreateFilter(args []interface{}) (filters.Filter, error) {
+	sargs, err := getStrings(args)
+	if err != nil {
+		return nil, err
+	}
+	if len(sargs) < 1 {
+		return nil, filters.ErrInvalidFilterParameters
+	}
+
+	issuerURL := sargs[0]
+
+	cfg, err := getOpenIDConfig(issuerURL)
+	if err != nil {
+		return nil, err
+	}
+
+	f := &jwtValidationFilter{
+		jwksUri: cfg.JwksURI,
+	}
+
+	return f, nil
+}
+
+func getKeyFunction(url string) (jwks *keyfunc.JWKs, err error) {
+	if jwks, ok := jwksMap[url]; ok {
+		return jwks, nil
+	}
+
+	// Create the keyfunc options. Use an error handler that logs. Refresh the JWKs when a JWT signed by an unknown KID
+	// is found or at the specified interval. Rate limit these refreshes. Timeout the initial JWKs refresh request after
+	// 10 seconds. This timeout is also used to create the initial context.Context for keyfunc.Get.
+	refreshInterval := time.Hour
+	refreshRateLimit := time.Minute * 5
+	refreshTimeout := time.Second * 10
+	refreshUnknownKID := true
+	options := keyfunc.Options{
+		RefreshErrorHandler: func(err error) {
+			log.Errorf("There was an error with the jwt.KeyFunc\nError:%s\n", err.Error())
+		},
+		RefreshInterval:   &refreshInterval,
+		RefreshRateLimit:  &refreshRateLimit,
+		RefreshTimeout:    &refreshTimeout,
+		RefreshUnknownKID: &refreshUnknownKID,
+	}
+
+	jwks, err = keyfunc.Get(url, options)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get the JWKs from the given URL %s Error:%s", url, err.Error())
+	}
+	jwksMap[url] = jwks
+	return jwks, nil
+
+}
+
+func (f *jwtValidationFilter) Request(ctx filters.FilterContext) {
+	r := ctx.Request()
+
+	var info tokenContainer
+	infoTemp, ok := ctx.StateBag()[oidcClaimsCacheKey]
+	if !ok {
+		token, ok := getToken(r)
+		if !ok || token == "" {
+			unauthorized(ctx, "", missingToken, "", "")
+			return
+		}
+
+		jwks, err := getKeyFunction(f.jwksUri)
+		if err != nil {
+			log.Fatalf("Failed to get the JWKs from the given URL %s. Error:%s", f.jwksUri, err.Error())
+			unauthorized(ctx, "", missingJWKs, "", "")
+			return
+		}
+
+		var claims jwt.MapClaims
+		parsedToken, err := jwt.ParseWithClaims(token, &claims, jwks.KeyFunc)
+		if err != nil {
+			log.Errorf("Error while parsing jwt token : %v.", err)
+			unauthorized(ctx, "", invalidToken, "", "")
+			return
+		} else if !parsedToken.Valid {
+			log.Errorf("Invalid token")
+			unauthorized(ctx, "", invalidToken, "", "")
+			return
+		} else if parsedToken.Header["alg"] == nil {
+			log.Errorf("alg must be defined")
+			unauthorized(ctx, "", invalidToken, "", "")
+			return
+		}
```

Comment:
I think jwt lib verifies presence of `alg` so this branch is unreachable
https://github.com/golang-jwt/jwt/blob/2bd8ee77fc3a6b7116c5b5121cd7d07324423a85/parser.go#L139-L145

We should be careful about `none` algorithm though and disable it explicitly, see https://auth0.com/blog/critical-vulnerabilities-in-json-web-token-libraries/

I think it is disabled by default https://github.com/golang-jwt/jwt#compliance but it is worth to verify/add a test.